Adds 'Mods only' visualizer

colinta opened the pull request:
I made this for my own purposes, but I'll offer it up in the spirit of open source.  Based on the Svelte visualizer, this one displays the current modifiers and nothing more.

I use it because I have a custom keyboard that implements sticky keys, and I needed to see them on screen (it's easy to lose track of which keys are pressed).

Not really useful for screencasting, but it does show how to create a new visualizer. 🤷‍♂️ 

Example of pressing "command+shift" and "command+control+alt+shift".

![screen shot 2019-02-27 at 11 08 05 am](https://user-images.githubusercontent.com/27570/53504478-0d9b0980-3a80-11e9-86e6-46638da17838.png)
![screen shot 2019-02-27 at 11 07 47 am](https://user-images.githubusercontent.com/27570/53504489-112e9080-3a80-11e9-85c5-53a08577a9da.png)

colinta committed:
adds 'Mods only' visualizer

colinta committed:
updates for Sonoma to ignore mouse events (resize frame on mods change)

colinta commented:
I updated this yesterday to include fixes for clicking "through" the window when no (or few) mods are visible. In sonoma, anyway, the mods window was swallowing clicks.

colinta commented:
Rad! Happy to make changes to it, it's a bit niche at the moment (it scratches an itch _I_ have, but ofc I'd like it to be useful to more). 

The default and Svelte plugins are useful and distinct (podcasts, pair programming) - this is more for personal use. But yeah improvements welcome. 👍

akitchen commented on keycastr/ModsVisualizer.h:
```diff
@@ -0,0 +1,61 @@
+//	Copyright (c) 2009 Stephen Deken
```
Please update source headers to reflect the last year you edited the file, and *your* name, since it's your code ;)

akitchen commented on keycastr/ModsVisualizer.h:
```diff
@@ -0,0 +1,61 @@
+
+@interface ModsVisualizerView : NSView
+{
+	uint32_t _flags;
```
Spaces, not tabs please

akitchen commented on keycastr/ModsVisualizer.h:
```diff
@@ -0,0 +1,61 @@
+	uint32_t _flags;
+}
+
+-(void) noteFlagsChanged:(uint32_t)flags;
```
Please reformat method declaration formatting and other usages to match more recent code from me, e.g. `- (void)noteFlagsChanged:(uint32_t)flags;`

akitchen reviewed (changes requested):
A licensing change and a couple minor formatting nits; I will look at more of the functional code locally and follow up later.

akitchen commented:
One other thing -- of course there's no real style guide or anything here, but at one point I thought it would make sense for the order of modifiers to match what Apple's macOS apps show in their menu items. Pretty sure it's control-option-shift-command.

colinta commented:
Good changes, I'll ship these soon. On Oct 19, 2023, at 4:16 PM, Andrew Kitchen ***@***.***> wrote:﻿
One other thing -- of course there's no real style guide or anything here, but at one point I thought it would make sense for the order of modifiers to match what Apple's macOS apps show in their menu items. Pretty sure it's control-option-shift-command.

—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you authored the thread.Message ID: ***@***.***>

colinta committed:
code style changes, and ordered modifiers '⌃⌥⇧⌘'

colinta commented:
How's that? 🙂

akitchen commented:
Overall this is fantastic to see and I can see how it could be useful. Thanks again.

Additional code feedback is mainly just about code formatting for now -- please use Xcode's default formatting for multi-line statements just to keep things as consistent as possible.

A few more thoughts:

- Could you make it remember its position on screen?
- How about adding some configuration options for font size, colors, and maybe fade out duration?
- I wonder if it would be possible (and useful) to display the 'fn' key as well?

colinta commented:
* Could you make it remember its position on screen?

I'm struggling with how easy this _should_ be vs how easy it _is_...

```objc
    [_visualizerWindow setFrameUsingName:@"mods visualizerFrame" force:YES];
    [_visualizerWindow setFrameAutosaveName:@"mods visualizerFrame"];
```
This used to "just work", but is having no effect now. Anything obvious I'm getting wrong?

```objc
    NSRect windowFrame = { MODS_WIDTH, 100, 0, 100 };
    _visualizerWindow = [[NSWindow alloc]
         initWithContentRect:windowFrame
        styleMask:NSWindowStyleMaskBorderless
        backing:NSBackingStoreBuffered
        defer:NO];
    [_visualizerWindow setLevel:NSScreenSaverWindowLevel];
    [_visualizerWindow setBackgroundColor:[NSColor clearColor]];
    [_visualizerWindow setMovableByWindowBackground:YES];
    [_visualizerWindow setFrameUsingName:@"mods visualizerFrame" force:YES];
    [_visualizerWindow setFrameAutosaveName:@"mods visualizerFrame"];
    [_visualizerWindow setOpaque:NO];

    _visualizerView = [[ModsVisualizerView alloc] init];
    [_visualizerWindow setContentView:_visualizerView];
    [_visualizerView noteFlagsChanged:0];
// ... below ...
- (void)noteFlagsChanged:(uint32_t)flags {
    [_visualizerView noteFlagsChanged:flags];
    NSRect r = _visualizerWindow.frame;
    CGFloat right = r.origin.x + r.size.width;
    r.size.width = _visualizerView.frame.size.width;
    r.origin.x = right - r.size.width;
    [_visualizerWindow setFrame:r display:NO];
}

// ModsVisualizerView
- (void)noteFlagsChanged:(uint32_t)flags {
    _flags = flags;
    NSRect frame = self.frame;
    frame.size.width = MODS_WIDTH / 4 * (CGFloat)[self flagsCount];
    self.frame = frame;
    [self setNeedsDisplay:YES];
}
```

colinta commented:
* I wonder if it would be possible (and useful) to display the 'fn' key as well?

<img width="532" alt="Screenshot 2023-11-02 at 12 10 27 PM" src="https://github.com/keycastr/keycastr/assets/27570/576c9aaf-b8b4-4b9e-ba84-a3806ae8e282">

colinta committed:
supports 'FN' key, and I had the order flipped on modifiers

colinta commented:
* How about adding some configuration options for font size, colors, and maybe fade out duration?

These... maybe I could use some help on these. 😬 

The rest are done, except I can't figure out the frame-auto-save feature.

colinta commented:
@akitchen I added you as a contributor to my fork, in case you wanted to do any more features/clean-up there.

colinta commented:
Let's not let another year go by! 😬 What's next? I tried again today to save/restore the frame using `autosave`, but had no luck. What changed? 🤔

akitchen commented:
I haven't debugged this or spent too much time looking at it, but your starting frame is a bit suspicious to me at first glance. Maybe use the NSMakeRect function to help create your initial frame and ensure the origin and size are set properly. This code is manipulating the frame of both the window and the view, which may or may not be correct. A frame's origin is in its parent view's coordinate system. Not sure if you're using a flipped coordinate system or not, but you have to account for that as well.

I hope it helps!

akitchen reviewed (changes requested):
'REQUEST_CHANGES' performed by Xcode.

Teufeuleu commented:
Hi, I just discovered Keycastr and I wanted to chime in here, because this is exactly the last feature I am missing!
I use apps (Cycling'74 Max, Ableton Live) on which I often use modifier key + click interactions.
The Svelte visualizer helps in this regard, but it's too big and obstructive as I don't need the modifier key placeholders. I just want something like the Default visualizer to display also modifier keys when they are pressed without combination with an "other" key.
I'm not sure how this is different from https://github.com/keycastr/keycastr/issues/284 but I feel like both would fill my need anyway.
So thank you all for maintaining and improving this awesome utility!

colinta committed:
giving up on autosave frame, using defaults directly

colinta committed:
if frame.origin.x is on left side of window, grow to the right. Otherwise grow to the left.

colinta commented:
Yup, I'm back. It's working as intended now, rebased on main, updated the code to grow the mods view according to its position on the current screen.

I avoided the autosave bug by simply not using it 🤷‍♂️ – it uses NSUserDefaults under the hood, so that's what I used, too.

akitchen commented:
Hey @colinta thank you for following up here, I'll pull this locally and give it a try.

akitchen commented:
This is really great to see - thanks again!

One thing I noticed, the second and subsequent times I run the app after enabling this visualizer, on startup I see a blank square until the first modifier key event comes through. If you could find a fix for that, I'd be happy taking this pretty much as-is and maybe doing some minor cleanups before an upcoming release.

Also, let me know your thoughts on maintaining this visualizer going forward. There are a lot of obvious enhancements that could be done as well as potential other issues, and any release of KeyCastr has to be as high quality as possible given limited development time and potentially long gaps between releases. 

Thanks again for your contribution! Much appreciated!

akitchen reviewed (changes requested):
Fix for phantom box on second and subsequent launch. TY!

akitchen commented:
Hey @colinta , I've been trying this out for a little while and I really like it! I'd like to figure out the remaining steps needed to include it in an upcoming release. Thanks again for sharing the work you put into this!

There are a few minor internal changes/improvements which I'll be happy to make, but I think we would need a few more user-facing things. Mainly color and size options, and the ability to include other keystrokes and even mouseclicks like the other visualizers.

I can help with some or all of that, but it also brings me to another topic. With the additional functionality to display regular keystrokes and mouse clicks it becomes more than just a "Mods" visualizer. Any thoughts on what it should be called?

colinta commented:
Awesome! Yeah funny after all these years, but it's my daily driver!I know we have "Svelte" but what about "Minimal"? "Chromeless"?On Oct 18, 2024, at 6:27 PM, Andrew Kitchen ***@***.***> wrote:﻿
Hey @colinta , I've been trying this out for a little while and I really like it! I'd like to figure out the remaining steps needed to include it in an upcoming release. Thanks again for sharing the work you put into this!
There are a few minor internal changes/improvements which I'll be happy to make, but I think we would need a few more user-facing things. Mainly color and size options, and the ability to include other keystrokes and even mouseclicks like the other visualizers.
I can help with some or all of that, but it also brings me to another topic. With the additional functionality to display regular keystrokes and mouse clicks it becomes more than just a "Mods" visualizer. Any thoughts on what it should be called?

—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you were mentioned.Message ID: ***@***.***>

SoCuul commented:
Any update on the status of this? 👀

akitchen commented:
Glad to see there's interest in this, I like it too :) I have a local branch with a handful of fixups and updates, but it will also need a bit more work before I'm comfortable including it in an official release. I hope to be able to share more of an update soon.

colinta commented:
I'm happy to help test it!

akitchen commented:
Hey @colinta , sorry this has dragged on for so long. I had your changes on a local branch where I did some cleanups and started to pave the way for the rest of the changes I think this needs in order to be included in the app, but haven't been able to make much progress on it for a while.

If you're still interested you can check out the [minimal-visualizer](https://github.com/keycastr/keycastr/tree/minimal-visualizer) branch and consider opening some new PRs to that branch after you've had a chance to test and play around with it for a bit. Adding a third visualizer plus the need to add more options to the mouse visualizer means other infrastructure work is needed in the app before this can be launched as a part of the app, but I'd still like to do it.

akitchen commented:
#330 is relevant to this

akitchen commented:
I changed the target branch for this to the minimal-visualizer branch. Please merge that into your local copy or open a new PR if that's easier. Thanks!

nicktgn commented:
Is this still planned to be merged ?

akitchen commented:
> Is this still planned to be merged ?

As is? No, but there's the minimal-visualizer branch which starts to prepare this for merging. It still needs a few things -- not much actually, but a settings pane and separate options from the other visualizers. Haven't had much time for it in a while.

Anyone who wants to contribute can open a PR into that branch

SoCuul commented:
I'll try to get this branch up to a state where it can be merged, but I just have a question about the about screen. Does it indeed load a view from a Quartz Compositor project?

I never got to actually use Quartz Compositor (I was 12 when it was officially marked as deprecated), but it's my understanding that loading of compositions might start to cause issues soon (wouldn't be surprised if it already does on Tahoe). Does the animation in the about screen matter enough to keep it in the project? It could likely be replaced with a Core Animation-based view these days.

SoCuul commented:
> > Is this still planned to be merged ?
> 
> As is? No, but there's the minimal-visualizer branch which starts to prepare this for merging. It still needs a few things -- not much actually, but a settings pane and separate options from the other visualizers. Haven't had much time for it in a while.
> 
> Anyone who wants to contribute can open a PR into that branch

@akitchen I've implemented everything mentioned above, plus a few more features in #341
Hope this visualizer can finally get merged!
